features/FEATURE_BLE/source/generic/FileSecurityDb.cpp

```diff
@@ -0,0 +1,381 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "FileSecurityDb.h"
+
+namespace ble {
+namespace generic {
+
+const uint16_t DB_VERSION = 1;
+
+#define DB_STORE_OFFSET_LOCAL_KEYS    (0)
+#define DB_STORE_OFFSET_PEER_KEYS     (DB_STORE_OFFSET_LOCAL_KEYS + sizeof(SecurityEntryKeys_t))
+#define DB_STORE_OFFSET_PEER_IDENTITY (DB_STORE_OFFSET_PEER_KEYS + sizeof(SecurityEntryKeys_t))
+#define DB_STORE_OFFSET_PEER_SIGNING  (DB_STORE_OFFSET_PEER_IDENTITY + sizeof(SecurityEntryIdentity_t))
+
+#define DB_STORE_OFFSET_LOCAL_KEYS_LTK  (DB_STORE_OFFSET_LOCAL_KEYS)
+#define DB_STORE_OFFSET_LOCAL_KEYS_EDIV (DB_STORE_OFFSET_LOCAL_KEYS_LTK + sizeof(ltk_t))
+#define DB_STORE_OFFSET_LOCAL_KEYS_RAND (DB_STORE_OFFSET_LOCAL_KEYS_EDIV + sizeof(ediv_t))
+
+#define DB_STORE_OFFSET_PEER_KEYS_LTK  (DB_STORE_OFFSET_PEER_KEYS)
+#define DB_STORE_OFFSET_PEER_KEYS_EDIV (DB_STORE_OFFSET_PEER_KEYS_LTK + sizeof(ltk_t))
+#define DB_STORE_OFFSET_PEER_KEYS_RAND (DB_STORE_OFFSET_PEER_KEYS_EDIV + sizeof(ediv_t))
+
+#define DB_STORE_OFFSET_PEER_IDENTITY_ADDRESS (DB_STORE_OFFSET_PEER_IDENTITY)
+#define DB_STORE_OFFSET_PEER_IDENTITY_IRK     (DB_STORE_OFFSET_PEER_IDENTITY + sizeof(address_t))
+#define DB_STORE_OFFSET_PEER_IDENTITY_ADDRESS_IS_PUBLIC (DB_STORE_OFFSET_PEER_IDENTITY_IRK + sizeof(irk_t))
+
+#define DB_STORE_OFFSET_PEER_SIGNING_COUNT (DB_STORE_OFFSET_PEER_SIGNING + sizeof(csrk_t))
+
+/* make size multiple of 4 */
+#define PAD4(value) ((((value - 1) / 4) * 4) + 4)
+
+#define DB_SIZE_STORE \
+    PAD4(sizeof(SecurityEntryKeys_t) + \
+    sizeof(SecurityEntryKeys_t) + \
+    sizeof(SecurityEntryIdentity_t) + \
+    sizeof(SecurityEntrySigning_t) + \
+    sizeof(sign_count_t))
+
+/* without the size of the file offset as we don't store it */
+#define DB_SIZE_ENTRY \
+    (sizeof(SecurityDistributionFlags_t) + sizeof(sign_count_t))
+
+#define DB_SIZE_ENTRIES \
+    (FileSecurityDb::MAX_ENTRIES * DB_SIZE_ENTRY)
+
+#define DB_SIZE_STORES \
+    (FileSecurityDb::MAX_ENTRIES * DB_SIZE_STORE)
+
+#define DB_OFFSET_VERSION          (0)
+#define DB_OFFSET_RESTORE          (DB_OFFSET_VERSION + sizeof(DB_VERSION))
+#define DB_OFFSET_LOCAL_IDENTITY   (DB_OFFSET_RESTORE + sizeof(bool))
+#define DB_OFFSET_LOCAL_CSRK       (DB_OFFSET_LOCAL_IDENTITY + sizeof(SecurityEntryIdentity_t))
+#define DB_OFFSET_LOCAL_SIGN_COUNT (DB_OFFSET_LOCAL_CSRK + sizeof(csrk_t))
+#define DB_OFFSET_ENTRIES          (DB_OFFSET_LOCAL_SIGN_COUNT + sizeof(sign_count_t))
+#define DB_OFFSET_STORES           (DB_OFFSET_ENTRIES + DB_SIZE_ENTRIES)
+#define DB_OFFSET_MAX              (DB_OFFSET_STORES + DB_SIZE_STORES)
+#define DB_SIZE                    PAD4(DB_OFFSET_MAX)
```

Comment:
Nice to have: have packed versions of these structures to avoid persisting the structures' padding bytes to flash

Reply:
Padding is there for easy blanking (simpler code), space taken up by file is unlikely to be affected. The whole file is likely to be smaller than a single page.